src/box/replication.cc

```diff
@@ -206,6 +210,71 @@ replicaset_set_sync_quorum(void)
 	replication_sync_quorum_auto = replicaset.applier.connected;
 }
 
+/**
+ * Deactivates all expired anonymous replicas. They are expected to be
+ * deleted on deactivation as orphan.
+ * Returns lowest time-to-live among remaining anonymous replicas via
+ * `lowest_ttl`. If there are no anonymous replicas left or the function has
+ * failed to deactivate a replica, `replication_anon_gc_timeout` is returned.
+ */
+static void
+replication_anon_gc(double *lowest_ttl)
+{
+	double timeout = replication_anon_gc_timeout;
+	while (true) {
+		*lowest_ttl = timeout;
+		struct replica *expired_replica = NULL;
+		/*
+		 * Search for the first expired replica. Note that we cannot
+		 * just deactivate it and continue the iteration because a yield
+		 * will happen and the replicaset can be changed significantly.
+		 */
+		replicaset_foreach(replica) {
+			if (!replica->anon || replica_has_connections(replica))
+				continue;
+			double last_seen = relay_last_row_time(replica->relay);
+			double unused = clock_monotonic() - last_seen;
+			double ttl = timeout - unused;
+			if (ttl <= 0) {
+				expired_replica = replica;
+				break;
+			}
```

Comment:
If I don't miss something, `lowest_ttl` is `anon_gc_timeout` in the case, however it is possible that there is a replica with a lower TTL in the tail of the list.

Reply:
That is why it is wrapped into a `while (true)` loop - it will stop only when there are no expired replicas left - then the whole list will be scanned and `lowest_ttl` will be correct.

Reply:
OK, got it.